// @mui-treasury/layout/EdgeSidebar/EdgeSidebar.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EdgeSidebar = exports.useLooseSidebarCtx = exports.useSidebarCtx = exports.SidebarContext = void 0;

var _react = _interopRequireWildcard(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _styles = require("@mui/material/styles");

var _Drawer = _interopRequireWildcard(require("@mui/material/Drawer"));

var _Root = require("../Root/Root");

var _useScreen = require("../hooks/useScreen");

var _pickNearestBreakpoint = require("../utils/pickNearestBreakpoint");

var _useSidebarAutoCollapse = require("../hooks/useSidebarAutoCollapse");

var _constant = require("../utils/constant");

var _EdgeOffset = require("./EdgeOffset");

var _WindowContext = require("../WindowContext");

var _excluded = ["children"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var SidebarContext = /*#__PURE__*/_react["default"].createContext(undefined);

exports.SidebarContext = SidebarContext;

var useSidebarCtx = function useSidebarCtx() {
  var ctx = (0, _react.useContext)(SidebarContext);

  if (!ctx) {
    throw new Error("useSidebarCtx must be called under <EdgeSidebar />");
  }

  return ctx;
};

exports.useSidebarCtx = useSidebarCtx;

var useLooseSidebarCtx = function useLooseSidebarCtx() {
  return (0, _react.useContext)(SidebarContext);
};

exports.useLooseSidebarCtx = useLooseSidebarCtx;

var hasAutoExpanded = function hasAutoExpanded(config) {
  return !!config && (typeof config.autoExpanded === "boolean" || typeof config.uncollapsedOnHover === "boolean");
};

var EdgeSidebarRoot = (0, _styles.styled)(_Drawer["default"], {
  name: "AppEdgeSidebar",
  slot: "Root",
  overridesResolver: function overridesResolver(props, styles) {
    return styles.root;
  }
})(function (_ref) {
  var ownerState = _ref.ownerState;
  return _objectSpread({}, (ownerState.entered || ownerState.variant === "permanent" || ownerState.variant === "persistent") && _defineProperty({}, "& .".concat(_Drawer.drawerClasses.paper), {
    transition: "".concat(_constant.CSS_TRANSITION, " !important"),
    transitionProperty: "all !important"
  }));
});

var EdgeSidebar = function EdgeSidebar(_ref3) {
  var _layoutState$sidebarI, _iDocument$body, _props$ModalProps, _props$PaperProps3, _props$PaperProps4, _props$PaperProps5, _builder$sidebarId;

  var children = _ref3.children,
      inProps = _objectWithoutProperties(_ref3, _excluded);

  var props = (0, _styles.useThemeProps)({
    props: inProps,
    name: "AppEdgeSidebar"
  });
  var anchor = props.anchor;

  if (!anchor) {
    throw new Error('Missing prop "anchor" on EdgeSidebar component');
  }

  var screen = (0, _useScreen.useScreen)();

  var _useWindowCtx = (0, _WindowContext.useWindowCtx)(),
      iDocument = _useWindowCtx.iDocument;

  var ctx = (0, _Root.useLayoutCtx)();
  var builder = ctx.builder,
      layoutState = ctx.state,
      setOpen = ctx.setOpen;
  var sidebarId = "".concat(anchor, "EdgeSidebar");
  var edgeSidebar = builder[sidebarId];
  var sidebarState = layoutState[sidebarId];
  var config = (0, _pickNearestBreakpoint.pickNearestBreakpoint)(edgeSidebar === null || edgeSidebar === void 0 ? void 0 : edgeSidebar.config, screen);
  (0, _useSidebarAutoCollapse.useSidebarAutoCollapse)(sidebarId); // auto-expanded feature

  var _React$useState = _react["default"].useState(false),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      entered = _React$useState2[0],
      setEntered = _React$useState2[1];

  var _React$useState3 = _react["default"].useState(false),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      expanded = _React$useState4[0],
      setExpanded = _React$useState4[1];

  var isMouseOverSidebar = _react["default"].useRef(false);

  var isAutoExpanded = hasAutoExpanded(config) && (config.autoExpanded || config.uncollapsedOnHover);

  var onMouseEnter = function onMouseEnter(event) {
    var _props$PaperProps, _props$PaperProps$onM;

    (_props$PaperProps = props.PaperProps) === null || _props$PaperProps === void 0 ? void 0 : (_props$PaperProps$onM = _props$PaperProps.onMouseEnter) === null || _props$PaperProps$onM === void 0 ? void 0 : _props$PaperProps$onM.call(_props$PaperProps, event);

    if (sidebarState !== null && sidebarState !== void 0 && sidebarState.collapsed && isAutoExpanded) {
      isMouseOverSidebar.current = true;
      setTimeout(function () {
        if (isMouseOverSidebar.current) {
          setExpanded(true);
        }
      }, _constant.EDGE_SIDEBAR_EXPAND_DELAY);
    }
  };

  var onMouseLeave = function onMouseLeave(event) {
    var _props$PaperProps2, _props$PaperProps2$on;

    (_props$PaperProps2 = props.PaperProps) === null || _props$PaperProps2 === void 0 ? void 0 : (_props$PaperProps2$on = _props$PaperProps2.onMouseLeave) === null || _props$PaperProps2$on === void 0 ? void 0 : _props$PaperProps2$on.call(_props$PaperProps2, event);
    isMouseOverSidebar.current = false;
    setExpanded(false);
  };

  var onEntered = function onEntered() {
    var _props$SlideProps, _props$SlideProps$onE;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    // @ts-ignore
    (_props$SlideProps = props.SlideProps) === null || _props$SlideProps === void 0 ? void 0 : (_props$SlideProps$onE = _props$SlideProps.onEntered) === null || _props$SlideProps$onE === void 0 ? void 0 : _props$SlideProps$onE.call.apply(_props$SlideProps$onE, [_props$SlideProps].concat(args));
    setEntered(true);
  };

  var onExit = function onExit() {
    var _props$SlideProps2, _props$SlideProps2$on;

    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    // @ts-ignore
    (_props$SlideProps2 = props.SlideProps) === null || _props$SlideProps2 === void 0 ? void 0 : (_props$SlideProps2$on = _props$SlideProps2.onExit) === null || _props$SlideProps2$on === void 0 ? void 0 : _props$SlideProps2$on.call.apply(_props$SlideProps2$on, [_props$SlideProps2].concat(args));
    setEntered(false);
  };

  _react["default"].useEffect(function () {
    if (!(sidebarState !== null && sidebarState !== void 0 && sidebarState.collapsed)) {
      setExpanded(false);
    }
  }, [sidebarState === null || sidebarState === void 0 ? void 0 : sidebarState.collapsed]);

  var sidebarValue = _react["default"].useMemo(function () {
    return {
      id: sidebarId,
      anchor: anchor,
      expanded: expanded,
      setExpanded: setExpanded
    };
  }, [sidebarId, anchor, expanded, setExpanded]);

  if (!screen) return null;
  if (!edgeSidebar || !edgeSidebar.id) return null;
  var responsiveVariant = edgeSidebar.getDrawerVariant();
  var variant = (0, _pickNearestBreakpoint.pickNearestBreakpoint)(responsiveVariant, screen);
  if (!variant) return null;

  var ownerState = _objectSpread(_objectSpread({}, props), {}, {
    entered: entered,
    variant: variant
  });

  return /*#__PURE__*/_react["default"].createElement(SidebarContext.Provider, {
    value: sidebarValue
  }, /*#__PURE__*/_react["default"].createElement(EdgeSidebarRoot, _extends({}, props, {
    open: (_layoutState$sidebarI = layoutState[sidebarId]) === null || _layoutState$sidebarI === void 0 ? void 0 : _layoutState$sidebarI.open,
    anchor: anchor,
    variant: variant,
    ownerState: ownerState,
    onClose: function onClose() {
      var _props$onClose;

      for (var _len3 = arguments.length, args = new Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
        args[_key3] = arguments[_key3];
      }

      (_props$onClose = props.onClose) === null || _props$onClose === void 0 ? void 0 : _props$onClose.call.apply(_props$onClose, [props].concat(args));
      setOpen(sidebarId, false);
    },
    SlideProps: _objectSpread(_objectSpread({}, props.SlideProps), {}, {
      onEntered: onEntered,
      onExit: onExit
    }),
    ModalProps: _objectSpread(_objectSpread({}, props.ModalProps), {}, {
      container: (_iDocument$body = iDocument === null || iDocument === void 0 ? void 0 : iDocument.body) !== null && _iDocument$body !== void 0 ? _iDocument$body : (_props$ModalProps = props.ModalProps) === null || _props$ModalProps === void 0 ? void 0 : _props$ModalProps.container
    }),
    PaperProps: _objectSpread(_objectSpread({}, props.PaperProps), {}, {
      className: (0, _clsx["default"])((_props$PaperProps3 = props.PaperProps) === null || _props$PaperProps3 === void 0 ? void 0 : _props$PaperProps3.className, expanded && "AppEdgeSidebar-paperExpanded"),
      style: _objectSpread(_objectSpread({}, (_props$PaperProps4 = props.PaperProps) === null || _props$PaperProps4 === void 0 ? void 0 : _props$PaperProps4.style), expanded && {
        width: config === null || config === void 0 ? void 0 : config.width,
        overflow: "visible"
      }),
      onMouseEnter: onMouseEnter,
      onMouseLeave: onMouseLeave,
      sx: _objectSpread(_objectSpread({}, (_props$PaperProps5 = props.PaperProps) === null || _props$PaperProps5 === void 0 ? void 0 : _props$PaperProps5.sx), (_builder$sidebarId = builder[sidebarId]) === null || _builder$sidebarId === void 0 ? void 0 : _builder$sidebarId.getSxProps())
    })
  }), variant && variant !== "temporary" && /*#__PURE__*/_react["default"].createElement(_EdgeOffset.EdgeOffset, {
    sidebarId: sidebarId
  }), typeof children === "function" ? children(_objectSpread(_objectSpread({}, ctx), {}, {
    expanded: expanded,
    entered: entered,
    isMouseOverSidebar: isMouseOverSidebar.current
  })) : children));
};

exports.EdgeSidebar = EdgeSidebar;
