// @mui-treasury/layout/Header/Header.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Header = exports.Subheader = exports.TopHeader = void 0;

var _react = _interopRequireDefault(require("react"));

var _AppBar = _interopRequireDefault(require("@mui/material/AppBar"));

var _styles = require("@mui/material/styles");

var _Root = require("../Root/Root");

var _constant = require("../utils/constant");

var _excluded = ["children"],
    _excluded2 = ["children"],
    _excluded3 = ["children"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var OffsetRoot = (0, _styles.styled)("div")({
  flexShrink: 0
});

var Offset = function Offset(_ref) {
  var _builder$element;

  var element = _ref.element;

  var _useLayoutCtx = (0, _Root.useLayoutCtx)(),
      builder = _useLayoutCtx.builder;

  return /*#__PURE__*/_react["default"].createElement(OffsetRoot, {
    className: "HeaderOffset",
    sx: {
      height: (_builder$element = builder[element]) === null || _builder$element === void 0 ? void 0 : _builder$element.getOffsetHeight()
    }
  });
};

var TopHeaderRoot = (0, _styles.styled)("div", {
  name: "AppTopHeader",
  slot: "Root",
  overridesResolver: function overridesResolver(props, styles) {
    return styles.root;
  }
})({
  zIndex: 1,
  transition: _constant.CSS_TRANSITION
});

var TopHeader = function TopHeader(_ref2) {
  var _builder$topHeader;

  var children = _ref2.children,
      inProps = _objectWithoutProperties(_ref2, _excluded);

  var props = (0, _styles.useThemeProps)({
    props: inProps,
    name: "AppTopHeader"
  });
  var ctx = (0, _Root.useLayoutCtx)();
  var builder = ctx.builder;
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(TopHeaderRoot, _extends({}, props, {
    sx: _objectSpread(_objectSpread({}, props.sx), (_builder$topHeader = builder.topHeader) === null || _builder$topHeader === void 0 ? void 0 : _builder$topHeader.getSxProps())
  }), typeof children === "function" ? children(ctx) : children), /*#__PURE__*/_react["default"].createElement(Offset, {
    element: "topHeader"
  }));
};

exports.TopHeader = TopHeader;
var SubheaderRoot = (0, _styles.styled)("div", {
  name: "AppSubheader",
  slot: "Root",
  overridesResolver: function overridesResolver(props, styles) {
    return styles.root;
  }
})({
  zIndex: 1,
  transition: _constant.CSS_TRANSITION
});

var Subheader = function Subheader(_ref3) {
  var _builder$subheader;

  var children = _ref3.children,
      inProps = _objectWithoutProperties(_ref3, _excluded2);

  var props = (0, _styles.useThemeProps)({
    props: inProps,
    name: "AppSubheader"
  });
  var ctx = (0, _Root.useLayoutCtx)();
  var builder = ctx.builder;
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(SubheaderRoot, _extends({}, props, {
    sx: _objectSpread(_objectSpread({}, props.sx), (_builder$subheader = builder.subheader) === null || _builder$subheader === void 0 ? void 0 : _builder$subheader.getSxProps())
  }), typeof children === "function" ? children(ctx) : children), /*#__PURE__*/_react["default"].createElement(Offset, {
    element: "subheader"
  }));
};

exports.Subheader = Subheader;
var HeaderRoot = (0, _styles.styled)(_AppBar["default"], {
  name: "AppHeader",
  slot: "Root",
  overridesResolver: function overridesResolver(props, styles) {
    return styles.root;
  }
})({
  transition: _constant.CSS_TRANSITION
});

var Header = function Header(_ref4) {
  var _builder$header;

  var children = _ref4.children,
      inProps = _objectWithoutProperties(_ref4, _excluded3);

  var props = (0, _styles.useThemeProps)({
    props: inProps,
    name: "AppHeader"
  });
  var ctx = (0, _Root.useLayoutCtx)();
  var builder = ctx.builder;
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(HeaderRoot, _extends({
    color: "default",
    elevation: 0
  }, props, {
    sx: _objectSpread(_objectSpread({}, props.sx), (_builder$header = builder.header) === null || _builder$header === void 0 ? void 0 : _builder$header.getSxProps())
  }), typeof children === "function" ? children(ctx) : children), /*#__PURE__*/_react["default"].createElement(Offset, {
    element: "header"
  }));
};

exports.Header = Header;
