// @mui-treasury/layout/Width/WidthModel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createWidthInterface = exports.sumExternalGap = exports.getCssWidth = void 0;

var _isNilOrEmpty = require("../utils/isNilOrEmpty");

var _toValidCssValue = require("../utils/toValidCssValue");

var getCssWidth = function getCssWidth(externalGap) {
  if (typeof externalGap === "string") {
    return "calc(100% - (".concat(externalGap, "))");
  }

  if (typeof externalGap === "number") {
    if (externalGap === 0) return "100%";
    return "calc(100% - ".concat(externalGap, "px)");
  }

  return "auto";
};

exports.getCssWidth = getCssWidth;

var sumExternalGap = function sumExternalGap(a, b) {
  if ((0, _isNilOrEmpty.isNilOrEmpty)(a) && (0, _isNilOrEmpty.isNilOrEmpty)(b)) return undefined;
  if (a === 0 && b === 0) return 0;
  if (!a && !(0, _isNilOrEmpty.isNilOrEmpty)(b)) return b;
  if (!(0, _isNilOrEmpty.isNilOrEmpty)(a) && !b) return a;
  return "".concat((0, _toValidCssValue.toValidCssValue)(a), " + ").concat((0, _toValidCssValue.toValidCssValue)(b));
};

exports.sumExternalGap = sumExternalGap;

var createWidthInterface = function createWidthInterface(externalGap) {
  return {
    value: externalGap,
    combine: function combine(w) {
      return createWidthInterface(sumExternalGap(externalGap, w.value));
    },
    getStyle: function getStyle() {
      return {
        width: getCssWidth(externalGap)
      };
    }
  };
};

exports.createWidthInterface = createWidthInterface;
