// @mui-treasury/layout/utils/constant.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_THEME = exports.CSS_TRANSITION = exports.EDGE_SIDEBAR_EXPAND_DELAY = exports.RIGHT_EDGE_SIDEBAR_ID = exports.LEFT_EDGE_SIDEBAR_ID = exports.FOOTER_ID = exports.CONTENT_ID = exports.HEADER_ID = void 0;

var _styles = require("@mui/material/styles");

var HEADER_ID = "header";
exports.HEADER_ID = HEADER_ID;
var CONTENT_ID = "content";
exports.CONTENT_ID = CONTENT_ID;
var FOOTER_ID = "footer";
exports.FOOTER_ID = FOOTER_ID;
var LEFT_EDGE_SIDEBAR_ID = "leftEdgeSidebar";
exports.LEFT_EDGE_SIDEBAR_ID = LEFT_EDGE_SIDEBAR_ID;
var RIGHT_EDGE_SIDEBAR_ID = "rightEdgeSidebar";
exports.RIGHT_EDGE_SIDEBAR_ID = RIGHT_EDGE_SIDEBAR_ID;
var EDGE_SIDEBAR_EXPAND_DELAY = 350;
exports.EDGE_SIDEBAR_EXPAND_DELAY = EDGE_SIDEBAR_EXPAND_DELAY;
var CSS_TRANSITION = "all 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms, color 0s";
exports.CSS_TRANSITION = CSS_TRANSITION;
var DEFAULT_THEME = (0, _styles.createTheme)();
exports.DEFAULT_THEME = DEFAULT_THEME;
