// @mui-treasury/layout/Fullscreen/Fullscreen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Fullscreen = void 0;

var _react = _interopRequireDefault(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _styles = require("@mui/material/styles");

var _FullscreenContext = _interopRequireDefault(require("./FullscreenContext"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var FullscreenRoot = (0, _styles.styled)("div", {
  name: "Fullscreen",
  slot: "Root"
})({
  height: "100vh",
  display: "flex",
  flexDirection: "column",
  overflow: "hidden"
});

var Fullscreen = function Fullscreen(props) {
  return /*#__PURE__*/_react["default"].createElement(_FullscreenContext["default"].Provider, {
    value: true
  }, /*#__PURE__*/_react["default"].createElement(FullscreenRoot, _extends({}, props, {
    className: (0, _clsx["default"])("Fullscreen", props.className)
  })));
};

exports.Fullscreen = Fullscreen;
